mmseg/datasets/pipelines/transforms.py

```diff
@@ -1040,3 +1042,259 @@ def __repr__(self):
         repr_str += f'fill_in={self.fill_in}, '
         repr_str += f'seg_fill_in={self.seg_fill_in})'
         return repr_str
+
+
+@PIPELINES.register_module()
+class RandomMosaic(object):
+    """Mosaic augmentation. Given 4 images, mosaic transform combines them into
+    one output image. The output image is composed of the parts from each sub-
+    image.
+
+    .. code:: text
+                        mosaic transform
+                           center_x
+                +------------------------------+
+                |       pad        |  pad      |
+                |      +-----------+           |
+                |      |           |           |
+                |      |  image1   |--------+  |
+                |      |           |        |  |
+                |      |           | image2 |  |
+     center_y   |----+-------------+-----------|
+                |    |   cropped   |           |
+                |pad |   image3    |  image4   |
+                |    |             |           |
+                +----|-------------+-----------+
+                     |             |
+                     +-------------+
+     The mosaic transform steps are as follows:
+         1. Choose the mosaic center as the intersections of 4 images
+         2. Get the left top image according to the index, and randomly
+            sample another 3 images from the custom dataset.
+         3. Sub image will be cropped if image is larger than mosaic patch
+    Args:
+        prob (float): mosaic probability.
+        img_scale (Sequence[int]): Image size after mosaic pipeline of single
+           image. Default: (640, 640).
```

Comment:
It's a little ambiguous. Maybe it's a good idea to modify it to:

Image size after mosaic pipeline of a single image. The size of the output image is four times that of a single image. The output image comprises 4 single images.

Reply:
I agree with your comment.
I think it would be better to change mosaic_combine like this image in top_left. (ex. random)

                           center_x
                +------------------------------+
                |       pad        |  pad      |
                |      +-----------+           |
                |      |           |           |
                |      |  image1   |--------+  |
                |      |           |        |  |
                |      |           | image2 |  |
     center_y   |----+-------------+-----------|
                |    |   cropped   |           |
                |pad |   image3    |  image4   |
                |    |             |           |
                +----|-------------+-----------+
                     |             |
                     +-------------+
                     
what do you think?

![image](https://user-images.githubusercontent.com/30465912/145686880-e184920e-16ad-4080-a60a-46055e8ebe78.png)

Reply:
another method:
`img_scale (Sequence[int])` -> `img_scale (tuple[int, int] | list[tuple[int, int]])` like cutout_shape

It seems easier to implement.


![image](https://user-images.githubusercontent.com/30465912/145713168-030bbdca-7ccc-4016-b714-f43741e3bf0a.png)

Reply:
Could you please explain pad1 and pad2?

Reply:
Do you mean different img_scale for different images (4 img_scale for 4 images)?

Reply:
Got it, we can random a scale ratio in L1163 like ratio range in `Resize`, the default ratio range is [0.5, 2.0] in mmseg.
Will this approach match your ideas better?
This is a more complicated and general mosaic augmentation method, but I am not sure it will work well for semantic segmentation (maybe we need to do some research or ablation).

Reply:
https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/datasets/pipelines/transforms.py#L96
https://github.com/open-mmlab/mmsegmentation/blob/bdfbaf00c1f55e054e6088e60f955345ab7383c0/mmseg/datasets/pipelines/transforms.py#L196

Reply:
> Got it, we can random a scale ratio in L1163 like ratio range in `Resize`, the default ratio range is [0.5, 2.0] in mmseg. Will this approach match your ideas better? This is a more complicated and general mosaic augmentation method,

I will check it.

>  but I am not sure it will work well for semantic segmentation (maybe we need to do some research or ablation).

As you said, it may not mean much.
Frist, in the current Mosaic, we will do ablation study. And then I will propose this method again in this PR or new PR.